src/Spice86.Core/Emulator/InterruptHandlers/Dos/DosInt21Handler.cs

```diff
@@ -411,7 +411,27 @@ public class DosInt21Handler : InterruptHandler {
     }
 
     public void PrintString() {
-        string str = GetDosString(_memory, _state.DS, _state.DX, '$');
+        ushort segment = _state.DS;
+        ushort offset = _state.DX;
+        string str = GetDosString(_memory, segment, offset, '$');
+
+        // Store registers
+        ushort[] savedRegisters = {_state.ES, _state.BP, _state.AX, _state.BX, _state.CX, _state.DX};
+        
+        // Set up values for bios call
+        _state.BL = 0x0F; // white on black
+        _state.ES = segment;
+        _state.BP = offset;
+        _state.CX = (ushort)str.Length;
+        _state.AL = 0x00; // no attribute, no cursor movement
+        _state.BH = _machine.Bios.CurrentVideoPage;
+        _state.DX = _machine.Bios.GetCursorPosition(_state.BH);
+        _state.AH = 0x13; // Write string
+        _machine.CallbackHandler.Run(0x10);
```

Comment:
Would be nice to have an accessible public method in video bios for this instead.

Reply:
I'm not sure. I made it this way so anything else can override / take over INT 10 and it'd still work. A hard dependency between DOS and the current VGA bios does not sound that beneficial.

Reply:
I suppose you are thinking about games that may provide their own implementation of INT 10 or loading a real VGA bios beside emulated DOS.
The thing is we are not able to call ASM code from C# (and it is a difficult problem because you don't have a guarantee the ASM will be nice and return to you).

Reply:
I have refactored the VgaBios to have a public interface usable by outside code.
And an interrupt interface that only handles interrupt functions.